### ISSUE TITLE
Customisable @-menu

### ISSUE THREAD

**Issue by mourningsun** [CleverRaven/Cataclysm-DDA#51156]
# Is your feature request related to a problem? Please describe.
Currently there is no way to switch off or edit UI elements on the character screen like you can do with the sidebar. I think the option to customise @-menu would benefit players who play on smaller screens or don't like the default look of the menu.

# Describe the solution you'd like
Add an option to customise @-menu.

# Describe alternatives you've considered
None

**Comment by Brambor** [comment]
Kind of related, about visible scrollbars: #50928

**Comment by Brambor** [comment]
In @: Currently, there are 9 windows in 3 columns, 8 of them are in a column, one spans all three columns.

Having 3 columns is different from the sidebar because there is only 1 column.

Do you, or anyone, have any mockups? How would you solve this 3-column dilemma?

I will add pictures from the last experimental shortly...

**Comment by mourningsun** [comment]
Maybe all 3 colums should have separate edit buttons?

**Comment by Brambor** [comment]
I'm thinking more like you can select number of columns and their width...

**Comment by Brambor** [comment]
Latest experimental with arbitrary size:

![image](https://user-images.githubusercontent.com/13402666/131225440-c23222ba-3783-428a-ae14-14ccad38068e.png)

Note: SKILLS tab don't have scrollbar unless selected, I will look into that...

**Comment by Brambor** [comment]
What's your resolution @mourningsun? If I am to look at it, I would like your eyes, not theoretical ones...

**Comment by mourningsun** [comment]
> What's your resolution @mourningsun? If I am to look at it, I would like your eyes, not theoretical ones...

I play on a tablet and I dont have a PC. My terminal size in game would be equivalent to 1045x576

**Comment by mourningsun** [comment]
> Latest experimental with arbitrary size:
> 
> ![image](https://user-images.githubusercontent.com/13402666/131225440-c23222ba-3783-428a-ae14-14ccad38068e.png)
> 
> Note: SKILLS tab don't have a scrollbar unless selected, I will look into that... Also, they are the only ones to have a scrollbar on the left.

What would happen if you remove one of the windows. Would they leave empty space or group together with the remaining windows?

**Comment by Brambor** [comment]
> What would happen if you remove one of the windows somewhere in the middle. Would they leave empty space or group together with the remaining windows?

Yes, that has to be addressed.
I think, when you make a multi-column window, it has to have windows in all columns above itself, or none (has to be first).
You still could get weird behaviour, but the player can see that and it would not crash, hopefully. They would just change it around.

So when you have a multi-column window, all windows above it get as much space, as the shortest column needs.
Then we just divide the space fairly, as I have done in #50892 for EFFECTS + PROFICIENCIES, only for N windows instead of just 2.

**Comment by Brambor** [comment]
I think, for now, If I could get a working version without multi-column and then add it after, it would be amazing still, right? Yeah, we will see. That's a nice project!

**Comment by Brambor** [comment]
I see, currently, the width of @ menu is decided like that, because it is the minimum width the game supports...

**Comment by mourningsun** [comment]
> I think, for now, If I could get a working version without multi-column and then add it after, it would be amazing still, right? Yeah, we will see. That's a nice project!
Cant wait to try it out!

**Comment by mourningsun** [comment]
> Latest experimental with arbitrary size:
> 
> ![image](https://user-images.githubusercontent.com/13402666/131225440-c23222ba-3783-428a-ae14-14ccad38068e.png)
> 
> Note: SKILLS tab don't have a scrollbar unless selected, I will look into that... Also, they are the only ones to have a scrollbar on the left.

The window in the middle with descriptions takes a lot of place, can it be smaller or maybe moved to the bottom of the screen?

**Comment by Brambor** [comment]
I hope you will be able to do all that in the future :)

**Comment by mourningsun** [comment]
Would there be separate editing screen or mode where you can rearrange and edit columns and their contents? I guess that would be hardest part to implement.

**Comment by Brambor** [comment]
> Would there be separate editing screen or mode where you can rearrange and edit columns and their contents? I guess that would be the hardest part to implement.

Yes, that's exactly what I would like to implement! Like it is with the sidebar, I thought you asked for that :D

**Comment by mourningsun** [comment]
> > Would there be separate editing screen or mode where you can rearrange and edit columns and their contents? I guess that would be the hardest part to implement.
> 
> Yes, that's exactly what I would like to implement! Like it is with the sidebar, I thought you asked for that :D

Of course!
 
Also here's how my character screen looks in my current game. I can post fonts and font settings later if necessary.

![Screenshot_2021-08-28-22-18-50](https://user-images.githubusercontent.com/74298401/131227237-74655614-f41e-42a6-90d2-6ea5ed8b0a37.png)
###

**Comment by Brambor** [comment]
Ok, there is what I would like to implement in the first PR from what you said:

> If there were an editing option I would:

- [x] stack Stats, Encumbrance and Speed on top of each other in the first column on the left
- [ ] and remove age, height, weight and blood type from Stats window.
- [x] Skills and Traits would take second column
- [x] and Effects and Proficiencies in third with hidden Bionics window.
- [x] The whole top part with a name would be hidden
- [x] and description window I would put at the bottom under all three columns or maybe also hide it.

> and remove age, height, weight and blood type from Stats window.

That's something for future improvement.

> The whole top part with a name would be hidden

Actually, Tere is not only name, but also `[y]Switch Gender    [?]`, this would probably fall under "Hint", like in the sidebar menu

> and description window I would put at the bottom under all three columns or maybe also hide it.

There is actually quite a lot going on, but that's the beauty of customization, your taste doesn't have to match with mine and we're both happy 😄

**Comment by Brambor** [comment]
I would add: The ability to remove titles (EFFECTS, TRAITS, BIONICS, SKILLS...), because experienced player doesn't need them.

**Comment by Night-Pryanik** [comment]
Closing as stale, since stalebot can't do this by itself.
